### PR TITLE
Convert byte_extract keyword parser to Rust

### PR DESCRIPTION
**Pull request by jlucovsky**
Continuation of #10881

Convert the byte_extract option parser from C to Rust.

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: [6873](https://redmine.openinfosecfoundation.org/issues/6873)

Describe changes:
- Refactor code in rust/src/detect to support re-usability
- Implement the byte_extract parser in Rust, with unittests
- Modify the existing byte_extract module to use the Rust parser

Updates:
- Apply rust formatting to byte_{extract,math}.rs with current version of rustfmt
- Use correct issue number in commit messages
- Use idiom `if let Some(v)` 

### Provide values to any of the below to override the defaults.

To use a pull request use a branch name like `pr/N` where `N` is the
pull request number.

Alternatively, `SV_BRANCH` may also be a link to an
OISF/suricata-verify pull-request.

```
SV_REPO=
SV_BRANCH=
SU_REPO=
SU_BRANCH=
LIBHTP_REPO=
LIBHTP_BRANCH=
```

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 20146

File: rust/src/detect/mod.rs [hunk 1]
```diff
@@ -43,6 +44,47 @@ pub trait EnumString<T> {
     fn from_str(s: &str) -> Option<Self> where Self: Sized;
 }
 
+#[repr(u8)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+// endian <big|little|dce>
+pub enum ByteEndian {
+    _EndianNone = 0,
```
Comment (catenacyber) [inline rust/src/detect/mod.rs:51]: Could you add a comment why we need this unused value ?

File: rust/src/detect/byte_extract.rs [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+                        "base specified before string; use \"string, base\"".to_string(),
+                    ));
+                }
+                byte_extract.flags |= DETECT_BYTE_EXTRACT_FLAG_STRING;
+            }
+            "oct" | "dec" | "hex" => {
+                if 0 == (byte_extract.flags & DETECT_BYTE_EXTRACT_FLAG_STRING) {
+                    return Err(make_error("string must be set first".to_string()));
+                }
+                if 0 != (byte_extract.flags & DETECT_BYTE_EXTRACT_FLAG_BASE) {
```
Comment (catenacyber) [inline rust/src/detect/byte_extract.rs:199]: This good code was added in an unrelated commit

File: rust/src/detect/byte_extract.rs [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+                if 0 != (byte_extract.flags & DETECT_BYTE_EXTRACT_FLAG_SLICE) {
+                    return Err(make_error("slice already set".to_string()));
+                }
+                byte_extract.flags |= DETECT_BYTE_EXTRACT_FLAG_SLICE;
+            }
+            "dce" | "big" | "little" => {
+                if 0 != (byte_extract.flags & DETECT_BYTE_EXTRACT_FLAG_ENDIAN) {
+                    return Err(make_error("endianess already set".to_string()));
+                }
+                byte_extract.endian = match get_endian_value(name) {
```
Comment (catenacyber) [inline rust/src/detect/byte_extract.rs:176]: Could we use `if let Some(endian) = get_endian_value(val) {` as is done in byte_math ?

File: rust/src/detect/mod.rs [hunk 4]
```diff
@@ -43,6 +44,47 @@ pub trait EnumString<T> {
...
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum ByteBase {
+    _BaseNone = 0,
+    BaseOct = 8,
+    BaseDec = 10,
+    BaseHex = 16,
+}
+
+fn get_string_value(value: &str) -> Option<ByteBase> {
+    let res = match value {
```
Comment (catenacyber) [inline rust/src/detect/mod.rs:67]: you can `return` without assigning to `res`

File: rust/src/detect/byte_extract.rs [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+pub const DETECT_BYTE_EXTRACT_FLAG_STRING: u16 = 0x02;
+pub const DETECT_BYTE_EXTRACT_FLAG_ALIGN: u16 = 0x04;
+pub const DETECT_BYTE_EXTRACT_FLAG_ENDIAN: u16 = 0x08;
+pub const DETECT_BYTE_EXTRACT_FLAG_SLICE: u16 = 0x10;
+pub const DETECT_BYTE_EXTRACT_FLAG_MULTIPLIER: u16 = 0x20;
+pub const DETECT_BYTE_EXTRACT_FLAG_NBYTES: u16 = 0x40;
+pub const DETECT_BYTE_EXTRACT_FLAG_OFFSET: u16 = 0x80;
+pub const DETECT_BYTE_EXTRACT_FLAG_BASE: u16 = 0x100;
+
+pub const DETECT_BYTE_EXTRACT_MULTIPLIER_DEFAULT: u16 = 1;
```
Comment (catenacyber) [inline rust/src/detect/byte_extract.rs:42]: This needs to be public only for C unit tests, right ?

File: rust/src/detect/byte_extract.rs [hunk 6]
```diff
@@ -0,0 +1,466 @@
...
+        SCDetectByteExtractData {
+            nbytes: values[0]
+                .parse::<u8>()
+                .map_err(|_| make_error(format!("invalid nbytes value: {}", values[0])))?,
+            ..Default::default()
+        }
+    };
+
+    let value = values[1]
+        .parse::<i32>()
```
Comment (catenacyber) [inline rust/src/detect/byte_extract.rs:120]: Why not parse a i16 directly ?

File: rust/src/detect/byte_extract.rs [hunk 7]
```diff
@@ -0,0 +1,466 @@
...
+            "invalid offset value: must be between {} and {}: {}",
+            i16::MIN,
+            i16::MAX,
+            value
+        )));
+    }
+
+    let value = values[2];
+    let (_, value) = parse_token(value)?;
+    match CString::new(value) {
```
Comment (catenacyber) [inline rust/src/detect/byte_extract.rs:135]: Could use `if let` instead of match

File: src/detect-byte-extract.c [hunk 8]
```diff
@@ -698,18 +423,13 @@ static int DetectByteExtractTest01(void)
 {
     int result = 0;
 
-    DetectByteExtractData *bed = DetectByteExtractParse(NULL, "4, 2, one");
+    SCDetectByteExtractData *bed = DetectByteExtractParse(NULL, "4, 2, one");
```
Comment (catenacyber) [inline src/detect-byte-extract.c:426]: Was not this unit test duplicated in rust ?
Can it go away from C ?

File: rust/src/detect/byte_extract.rs [hunk 9]
```diff
@@ -0,0 +1,466 @@
...
+        }
+    }
+
+    for value in values.iter().skip(DETECT_BYTE_EXTRACT_FIXED_PARAM_COUNT) {
+        let (mut val, mut name) = take_until_whitespace(value)?;
+        val = val.trim();
+        name = name.trim();
+        match name {
+            "align" => {
+                if 0 != (byte_extract.flags & DETECT_BYTE_EXTRACT_FLAG_ALIGN) {
```
Comment (catenacyber) [inline rust/src/detect/byte_extract.rs:152]: Still wonder how to improve this pattern.

Maybe like `check_flags(&byte_extract.flags, DETECT_BYTE_EXTRACT_FLAG_ALIGN, "align")?`

**Review by catenacyber — CHANGES_REQUESTED** [review]
Thanks for the work Jeff

Just some nits left for me

CI : 🟢 
Code : Checking now
Commits segmentation : Ok like that, or also ok to squash last commit... Also looks like some rustfmt changes could be run first in parse_bytemath tests
Commit messages : good
Git ID set : looks fine for me
CLA : :-p
Doc update : not needed
Redmine ticket : ok
Rustfmt : Could you run it to have 2 small changes in byte_extract.rs and byte_math.rs ?
Tests : ok
Dependencies added: none added

File: rust/src/detect/byte_extract.rs (hunk 3 above)
Comment (jlucovsky) [reply rust/src/detect/byte_extract.rs:176]: Yes; will do.

File: src/detect-byte-extract.c [hunk 10]
```diff
@@ -211,255 +185,20 @@ int DetectByteExtractDoMatch(DetectEngineThreadCtx *det_ctx, const SigMatchData
...
+    SCDetectByteExtractData *bed = SCByteExtractParse(arg);
+    if (bed == NULL) {
+        SCLogError("invalid byte_extract values");
         goto error;
     }
 
-    bed = SCCalloc(1, sizeof(DetectByteExtractData));
-    if (unlikely(bed == NULL))
+    if (bed->flags & DETECT_BYTE_EXTRACT_FLAG_SLICE) {
+        SCLogError("byte_extract slice not yet supported");
```
[the remark this answers is not in the document]
Comment (jlucovsky) [reply src/detect-byte-extract.c:198]: Yes; will do.

File: rust/src/detect/mod.rs (hunk 1 above)
Comment (jlucovsky) [reply rust/src/detect/mod.rs:51]: It's not needed; will remove.

File: rust/src/detect/mod.rs (hunk 4 above)
Comment (jlucovsky) [reply rust/src/detect/mod.rs:67]: will do.

File: rust/src/detect/byte_extract.rs (hunk 6 above)
Comment (jlucovsky) [reply rust/src/detect/byte_extract.rs:120]: WIll do.

File: src/detect-byte-extract.c (hunk 8 above)
Comment (jlucovsky) [reply src/detect-byte-extract.c:426]: There's little benefit, imo, to removing existing test cases. There are rust based unittests but those are not exercised with `suricata -u`

File: rust/src/detect/byte_extract.rs (hunk 7 above)
Comment (jlucovsky) [reply rust/src/detect/byte_extract.rs:135]: will do.

File: rust/src/detect/byte_extract.rs (hunk 5 above)
Comment (catenacyber) [reply rust/src/detect/byte_extract.rs:42]: Sure `DETECT_BYTE_EXTRACT_MULTIPLIER_DEFAULT` is needed, but the `pub` is only needed for C unit tests, right ?

File: src/detect-byte-extract.c (hunk 8 above)
Comment (catenacyber) [reply src/detect-byte-extract.c:426]: I think both rust and C unit tests are run with `make check`

> There's little benefit, imo, to removing existing test cases. 

I would not say it had been removed, just moved ;-)

**Comment by jlucovsky** [comment]
Continued in #11017
